keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,837 @@
+
+<<[UNRESOLVED where to configure the cluster wide ID mappings ]>>
+
+#### Option 1: Configure in Kubelet Configuration File
+
+The ID mappings used for pods in `Cluster` mode are set in the kubelet
+configuration file. The file allows to set different ID mappings for user and
+group. Kubelet performs a check looking for configuration mistakes, like
+overlapping mappings, to prevent kubelet sending wrong ID mappings to the
+runtime.
+
+The following is an example a configuration file:
+
+```
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+clusterIDMappings:
+  uidMappings:
+  - containerID: 0
+    hostID: 500000
+    size: 65536
+  - containerID: 100000
+    hostID: 600000
+    size: 65536
+  gidMappings:
+  - containerID: 0
+    hostID: 500000
+    size: 65536
+
+```
+
+**Pros**:
+ - It's directly accessible by the kubelet. It's not needed to expose through an
+   API and there are not concerns if the api server is down.
+ - No modifications are needed to other components.
+
+**Cons**:
+ - It's more difficult for the user as they have to guarantee the same ID
+   mapping is configured in all the nodes.
+
+#### Option 2: Configure as a Cluster Parameter in kube-apiserver
+
+This option considers setting this parameter on the kube-apiserver.
+
+**Pros**:
+ - Easier for the user as the parameter is configured once in a single place.
+
+ **Cons**:
+ - It's difficult to expose this parameter to the kubelet.
+ - The parameter could not be available for the kubelet if the kube-apiserver is
+   down.
+
+<<[/UNRESOLVED]>>
+
+### 1-to-1 Mapping for fsGroup
+
+The `fsGroup` is asssumed to be the correct group owner of the files present on
+the volumes. A 1-to-1 ID mapping is added to ensure that the same GID used in
+the container is the effective GID in the host when the fsGroup is defined. If
+the `fsGroup` is part of the cluster ID mappings, a hole is "punched", otherwise
+an extra one element ID mapping is added.
+
+For instance, if the cluster GID mappings are
+
+```
+ContainerID HostID Size
+0           1000   0
+```
+and there is a pod with `fsGroup: 500`, the GID mappings for that specific pod
+are
+
+```
+ContainerID HostID Size
+0           1000   500
+501         1501   499
+500         500    1
+```
+
+On the other hand, if a pod has `fsGroup: 3000`, the GID mappings for that pod are
+
+```
+ContainerID HostID Size
+0           1000   1000
+3000        3000   1
+```
+
+### Updating Ownership of Ephemeral Volumes
+
+Ephemeral volumes use
+[`AtomicWriter`](https://github.com/kinvolk/kubernetes/blob/master/pkg/volume/util/atomic_writer.go)
+to create the files that are mounted to the containers. This component [has some
+logic](https://github.com/kinvolk/kubernetes/blob/c94242a7b1d238cc27aea9b6d45ccb9963e814bb/pkg/volume/util/atomic_writer.go#L403)
+to update the ownership of those files in some cases. It will be extended to
+take the ID mappings into consideration when the pod runs in `Cluster` mode.
+
+### Test Plan
+
```

Comment:
I know this is still in progess and provisional, but wanted you to know that once you move to implementable for a certain release (1.2x)  test plans will be required.

Reply:
Thanks!. Yes, I'd like to discuss first the technical details of the KEP and then I'll work on the test plan.

Reply:
sounds perfect :+1: